File: src/main/resources/templates/Application.java
package {{packageName}};

import org.springframework.boot.SpringApplication;

{{#isDubboClient}}
import javax.annotation.PostConstruct;
import com.alibaba.dubbo.config.annotation.Reference;
import {{dubboServiceName}};
{{/isDubboClient}}
{{^hasWeb}}
import java.util.concurrent.CountDownLatch;
{{/hasWeb}}


{{applicationImports}}
{{applicationAnnotations}}
public class {{applicationName}} {

	public static void main(String[] args) {{^hasWeb}}throws InterruptedException{{/hasWeb}}{

		SpringApplication.run({{applicationName}}.class, args);

		{{^hasWeb}}
		new CountDownLatch(1).await(); //hold住应用，防止provider退出
		{{/hasWeb}}
	}
	
}

File: src/main/resources/templates/Application.kt
package {{packageName}}

{{^kotlinSupport}}
import org.springframework.boot.SpringApplication
{{/kotlinSupport}}
{{applicationImports}}
{{#kotlinSupport}}
import org.springframework.boot.runApplication
{{/kotlinSupport}}

{{#isDubboClient}}
import javax.annotation.PostConstruct;
import com.alibaba.dubbo.config.annotation.Reference;
import {{dubboServiceName}};
{{/isDubboClient}}

{{applicationAnnotations}}
class {{applicationName}} {
    {{#isDubboClient}}
    @Reference(version = "{{dubboServiceVersion}}"{{^embeddedZookeeper}}, url = "dubbo://localhost:20880"{{/embeddedZookeeper}})
    var demoService: {{serviceSimpleName}}? = null;
    {{/isDubboClient}}

    {{#isDubboClient}}
    @PostConstruct
    fun init() {
        val sayHello = demoService!!.sayHello("world")
        System.err.println(sayHello)
    }
    {{/isDubboClient}}
}

fun main(args: Array<String>) {
{{^kotlinSupport}}
    {{#isDubboServer}}{{#embeddedZookeeper}}
    // start embedded zookeeper server
    EmbeddedZooKeeper(2181, false).start();

    {{/embeddedZookeeper}}{{/isDubboServer}}
    SpringApplication.run({{applicationName}}::class.java, *args)
{{/kotlinSupport}}
{{#kotlinSupport}}
    runApplication<{{applicationName}}>(*args)
{{/kotlinSupport}}
}